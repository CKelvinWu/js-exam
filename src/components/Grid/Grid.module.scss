.grid {
  display: flex;
}

.item {
  position: relative;
  width: 100%;
  height: 100%;
  overflow: auto;
}

.label {
  position: absolute;
  top: 5px;
  right: 17px;
  z-index: 10;
  padding: 2px 12px 2px 8px;
  color: white;
  background-color: rgba($color: #000000, $alpha: 0.5);
}
