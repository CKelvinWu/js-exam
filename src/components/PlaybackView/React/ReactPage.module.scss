.app {
  display: flex;
  flex-direction: row;
}

.rightWidget {
  height: 33.3%;
  width: 100%;
  border: 1px solid #eee;
}

.label {
	position: absolute;
  margin-top: 5px;
  right: 17px;
	z-index: 10;
	padding: 2px 12px;
	color: white;
	background-color: rgba($color: #000000, $alpha: 0.5);
}