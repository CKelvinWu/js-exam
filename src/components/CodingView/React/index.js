import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import 'brace';
import 'brace/mode/jsx';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import Grid from 'components/Grid';
import GridItem from 'components/Grid/GridItem';
import ConsoleWidget from 'components/Widgets/ConsoleWidget';
import CodeWidget from 'components/Widgets/CodeWidget';
import ResultWidget from 'components/Widgets/ResultWidget';
import AnswerWidget from 'components/Widgets/AnswerWidget';

import debouncedRunCode from 'utils/runCode';
import { REACT as GRID_LABEL_REACT } from 'utils/gridLabel';

import styles from './ReactPage.module.scss';

class ReactPage extends Component {
  static propTypes = {
    isExaming: PropTypes.bool,
    code: PropTypes.string,
    compiledCode: PropTypes.string,
    consoleMsg: PropTypes.array,
    handleCodeChange: PropTypes.func,
    wrappedConsole: PropTypes.object,
    resetConsole: PropTypes.func,
  };

  controlHeight = 70;

  componentDidMount() {
    const { compiledCode, wrappedConsole, resetConsole } = this.props;
    resetConsole();
    debouncedRunCode({ code: compiledCode, wrappedConsole });
  }

  shouldComponentUpdate(nextProps) {
    const { compiledCode: previousCompiledCode } = this.props;
    const { compiledCode, wrappedConsole, resetConsole } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      resetConsole();
      // WARNING: This is not debounced
      debouncedRunCode({ code: compiledCode, wrappedConsole });
    }
    return true;
  }

  render() {
    const { isExaming, code, handleCodeChange, consoleMsg } = this.props;
    const layout = [
      {
        key: 'code',
        x: 0,
        y: 0,
        width: window.innerWidth / 2 + 50,
        height: window.innerHeight / 2 + 100,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
      {
        key: 'answer',
        x: 0,
        y: 1,
        width: window.innerWidth / 2 + 50,
        height: window.innerHeight / 2 - 150,
        minWidth: 100,
        maxWidth: 700,
      },
      {
        key: 'result',
        x: 1,
        y: 0,
        width: window.innerWidth / 2 - 50,
        height: (window.innerHeight - this.controlHeight) / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
      {
        key: 'console',
        x: 1,
        y: 1,
        width: window.innerWidth / 2 - 50,
        height: (window.innerHeight - this.controlHeight) / 2 - 50,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
    ];
    return (
      <div className={styles.app}>
        <Grid layout={layout} totalWidth="100%" totalHeight="100%" autoResize>
          <GridItem key="code" label={GRID_LABEL_REACT.code}>
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              readOnly={!isExaming}
              mode="jsx"
              theme="monokai"
            />
          </GridItem>
          <GridItem key="answer" label={GRID_LABEL_REACT.answer}>
            {isExaming ? <AnswerWidget /> : null}
          </GridItem>
          <GridItem key="result" label={GRID_LABEL_REACT.result}>
            {isExaming ? <ResultWidget /> : null}
          </GridItem>
          <GridItem key="console" label={GRID_LABEL_REACT.console}>
            {isExaming ? <ConsoleWidget data={consoleMsg} /> : null}
          </GridItem>
        </Grid>
      </div>
    );
  }
}

export default withRouter(ReactPage);
