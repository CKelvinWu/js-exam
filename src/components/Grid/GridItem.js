import React from 'react';
import styles from './Grid.module.scss';

const GridItem = ({ label, children }) => (
  <div className={styles.item}>
    {label && <div className={styles.label}>{label}</div>}
    {children}
  </div>
);

export default GridItem;
