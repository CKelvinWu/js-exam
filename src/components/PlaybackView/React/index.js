import React, { Component } from 'react';
import PropTypes from 'prop-types';
import 'brace';
import 'brace/mode/jsx';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import { Spin } from 'antd';

import CodeWidget from 'components/Widgets/CodeWidget';
import ResultWidget from 'components/Widgets/ResultWidget';
import AnswerWidget from 'components/Widgets/AnswerWidget';

import debouncedRunCode from 'utils/runCode';
import { REACT as GRID_LABEL_REACT } from 'utils/gridLabel';

import styles from './ReactPage.module.scss';

class ReactPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      height: props.height,
      widthCode: props.width,
      widthTests: props.width,
    };
  }

  async componentDidMount() {
    const { compiledCode, addTape } = this.props;

    this.setState({
      height: `${window.innerHeight - 160}px`,
      widthCode: `${window.innerWidth * 0.45}px`,
      widthTests: `${window.innerWidth * 0.3}px`,
    });
    debouncedRunCode({ code: compiledCode, onTapeUpdate: addTape });
  }

  shouldComponentUpdate(nextProps) {
    
    const { compiledCode: previousCompiledCode } = this.props;
    const { compiledCode } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      debouncedRunCode({ code: compiledCode });
    }
    return true;
  }

  render() {
    this.codeWidgetstyle = {
      height: this.state.height,
      width: this.state.widthCode,
      position: 'relative',
    };
    this.testWidgetStyle = {
      height: this.state.height,
      width: this.state.widthTests,
      position: 'relative',
    };
    const { handleCodeChange, test, code } = this.props;
    return (
      <React.Fragment>
        <div className={styles.app}>
          <div style={this.codeWidgetstyle}>
            <div className={styles.label}>Code</div>
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              mode="jsx"
              theme="monokai"
              readOnly
            />
          </div>
          <div style={this.testWidgetStyle}>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.test}</div>
              <CodeWidget data={test} mode="jsx" theme="textmate" readOnly />
            </div>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.answer}</div>
              <AnswerWidget />
            </div>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.result}</div>
              <ResultWidget />
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

ReactPage.propTypes = {
  compiledCode: PropTypes.string,
  code: PropTypes.string,
  height: PropTypes.string,
  width: PropTypes.string,
  test: PropTypes.string,
  handleCodeChange: PropTypes.func,
  addTape: PropTypes.func,
};

export default ReactPage;
