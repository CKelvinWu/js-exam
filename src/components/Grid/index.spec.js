import React from 'react';
import { mount } from 'enzyme';
import Grid from './index';

describe('Grid index spec', () => {
  const mockLayout = [
    {
      key: 'code',
      x: 0,
      y: 0,
      width: 647,
      height: 462,
      minWidth: 100,
      minHeight: 100,
      maxWidth: 700,
      maxHeight: 924,
    },
    {
      key: 'test',
      x: 0,
      y: 1,
      width: 647,
      height: 462,
      minWidth: 100,
      maxHeight: 700,
    },
    {
      key: 'tape',
      x: 1,
      y: 0,
      width: 647,
      height: 924,
      minWidth: 100,
      minHeight: 100,
      maxWidth: 700,
      maxHeight: 500,
    },
  ];
  it('should render Grid properly', () => {
    const element = mount(
      <Grid
        layout={mockLayout}
        totalWidth="100%"
        totalHeight="100%"
        autoResize
        borderSize={1}
      >
        {[<div key={0}>test</div>, <div key={1}>test2</div>]}
      </Grid>,
    );
    expect(element.render()).toMatchSnapshot();
  });
});
