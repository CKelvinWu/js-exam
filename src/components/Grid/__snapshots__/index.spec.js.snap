// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid index spec should render Grid properly 1`] = `
<div
  class="grid"
  style="width: 100%; height: 100%;"
>
  <div
    class="border  horizontal"
    style="width: 648px; padding-right: 1px;"
  >
    <div
      class="border  vertical"
      style="width: 647px; height: 462px; padding-bottom: 1px;"
    />
    <div
      class="border "
      style="width: 647px; height: -462px;"
    />
  </div>
  <div
    class="border "
    style="width: -647px;"
  >
    <div
      class="border "
      style="width: -647px; height: 0px;"
    />
  </div>
</div>
`;
